whelktool/scripts/cleanups/2022/02/lxl-3817-clean-TopicSubdivisions.groovy

```diff
@@ -0,0 +1,104 @@
+/*
+
+Examples of
+rule 1:
+[(Alexander den store), (i litteraturen)] => (Alexander den store), (Motiv i litteraturen)
+rule 2:
+[(Symboler) + (i konsten)] => (Symboler i konsten); the combination exists
+rule 3:
+[(Kommunikation) + (i konsten)] => (Kommunikation); the combination does not exist
+rule 4:
+The rest.
+> 2 termComponents or combinations of subdivision and local entities.
+
+See https://jira.kb.se/browse/LXL-3817 for details.
+
+*/
+
+import whelk.Document
+import whelk.util.DocumentUtil
+
+SUBDIVISION_TO_SUBJECT = [
+        "https://id.kb.se/term/sao/i%20Bibeln"      : "https://id.kb.se/term/sao/Bibliska%20motiv",
+        "https://id.kb.se/term/sao/i%20filmen"      : "https://id.kb.se/term/sao/Motiv%20i%20filmen",
+        "https://id.kb.se/term/sao/i%20konsten"     : "https://id.kb.se/term/sao/Motiv%20i%20konsten",
+        "https://id.kb.se/term/sao/i%20litteraturen": "https://id.kb.se/term/sao/Motiv%20i%20litteraturen",
+        "https://id.kb.se/term/sao/i%20massmedia"   : "https://id.kb.se/term/sao/Motiv%20i%20massmedia",
+        "https://id.kb.se/term/sao/i%20pressen"     : "https://id.kb.se/term/sao/Motiv%20i%20pressen",
+]
+
+selectByCollection('bib') { bib ->
+    def data = bib.doc.data
+    DocumentUtil.traverse(data, { value, path ->
+        if (!(value instanceof Map && value.'@type'.equals("ComplexSubject") && value.termComponentList)) {
+            return
+        }
+
+        def components = value.termComponentList
+        def uriSubdivision = components.find { isSubdivision(it)}?.'@id'
+        def uri = components.find { !isSubdivision(it) && it.'@id' }?.'@id'
+
+        if (components.size() == 2 && uriSubdivision && uri) {
+            boolean rule1 = false
+            boolean rule2 = false
+            boolean rule3 = false
+
+            def prefLabelForSubdivision
+            def prefLabelForSubject
+
+            selectByIds([uriSubdivision]) { d ->
+                Map thing = d.doc.data['@graph'][1]
+                prefLabelForSubdivision = thing.prefLabel
+            }
+
+            selectByIds([uri]) { d ->
+                Map thing = d.doc.data['@graph'][1]
+                prefLabelForSubject = thing.prefLabel
+                if (thing.'@type' in ["Person", "Geographic", "Organization"]) {
```

Comment:
`Jurisdiction`, `Meeting` and `Family` are also subclasses of `Agent` too, maybe add these to the list?

Reply:
The reason they were not included was that there probably aren't many cases involving those agents. But I'll add them!